README.md
# mic_stream: 0.3.0

[Flutter Plugin]
Provides a tool to get the microphone input as 8 or 16 bit PCM Stream

## About mic_stream:

As Flutter still lacks some functionality, this plugin aims to provide the possibility to easily get an audio stream from the microphone, using a simple java implementation [=> Android only]

## How to use:

### Permissions

Make sure you have microphone recording permissions enabled for your project.
To do so, add this line to the AndroidManifest.xml:

`<uses-permission android:name="android.permission.RECORD_AUDIO"/>`

The plugin provides one class:

`Microphone <T> ({options})`

`{options}` are as described on the [Android page](https://developer.android.com/reference/android/media/AudioRecord#AudioRecord(int,%20int,%20int,%20int,%20int).
Use the provided enums instead of raw integers.

`T` is a generic type describing what type of stream you get, i.e., the stream will be of type `Stream <T>`.
It must be either `List <int>` or `int`; due to dart's dynamic typing system, supporting more is unreasonable.
`int` is the preferable choice, `List <int>` is the more raw variant supported for legacy reasons.

Call `.stream` on it to get a single-subscription stream or `.asBroadcast` to get a broadcast stream.
While someone listens to either of these streams, audio is being recorded;
stopping a subscription will cancel the recording.

NOTE: The stream will listen indefinitely if not cancelled otherwise

#### Example:

```
// Create a new stream
Microphone <int> mic = Microphone <int> (sampleRate: 44100);

// Start listening to the stream
StreamSubscription <int> listener = mic.stream.listen((samples) => print(samples));
```
...
```
// Cancel the subscription
listener.cancel()
```

### Legacy

To support backwards compatibility, the old function
```microphone({options})```
will be continued but its implementation is changed to use a static Microphone object instead.

## Flutter

About Flutter Plugins:
https://flutter.io/developing-packages/

Flutter Documentation:
https://flutter.io/docs


CHANGELOG.md
## 0.3.0

* Rewrite the whole setup
  * Provide Microphone class taking same input parameters
    * Provides `.stream` and `.asBroadcast` getters to get the audio stream
* Rewrite `microphone` function to use above Microphone class instead
  * Will be maintained for backwards compatibility

### 0.2.1

* Update permission plugin dependency

* Make AndroidX compatible


#### 0.2.0+2

* Formatted Code.


#### 0.2.0+1

* Updated README.


### 0.2.0

* Fixed value truncation, now ensuring correct values for 16BIT_PCM

+ Added a better example application in ./example/lib/main.dart


### 0.1.5

* Fixed Issue #8, causing immediate crashes in the latest flutter update


### 0.1.4

- Removed Debug output

* Fixed casting issue

+ Added Error handling for exceptions thrown on wrong AudioRecord params


### 0.1.3

* Updated README.


### 0.1.2

+ Added 16Bit PCM mode


### 0.1.1

+ Added customisability for the recorder
+ Added some error handling of the input params


## 0.1.0

* Rewritten Plugin to make use of EventChannel.StreamHandlers (Much nicer than before)

* microphone({Options}) returns a Stream<List<int>>
  * The stream starts upon onListen() and runs until onCancel()

* listen to the stream to start recording
* stop the subscription to stop


### 0.0.8

- Calculations of durations, as multi-threading currently makes it not working. Will be included in future releases

+ Some unit tests (Will be extended in the future to fully ensure the plugin's working)


### 0.0.7

* Fixed crucial bug from 0.0.6


### 0.0.6

* Smaller changes


### 0.0.5

+ Setter for microphone (audioSource, sampleRate, channelConfig, audioFormat)


### 0.0.4

* Changed Values for default and maximum sample rate (to 32 kHZ and 48 kHZ, respectively)


### 0.0.3

+ pause:            Pauses writing data to the stream
+ resume:           Resumes a paused stream


### 0.0.2

+ Getter for internal stream: microphone.stream


### 0.0.1

Initial release - Android support only!

Provides the Class Microphone inheriting StreamController<T> with .start() and .stop() to start and stop sending an audio stream from the microphone to the Microphone's internal Stream (Microphone.stream).
Also provides Microphone,broadcast which works the same but allows multiple StreamSubscriptions.
On start, a timestamp is set and returned on stop.

Constructor takes the Sample Rate as optional argument.

##### Provides methods to:
* platformVersion:  Getter to return current platform version
* bufferSize:       Getter to return current buffer size (calculated from the sample size)
* isRecording:      Returns the state of the class
* sampleRate:       Setter to manually update the sample rate (use with caution, though)
* close:            Closes the internal StreamController
* toString:         Pass-through to internal StreamController



lib/mic_stream.dart
import 'dart:async';

import 'package:permission/permission.dart';
import 'package:flutter/services.dart';

import 'package:mic_stream/microphone/Microphone.dart';

// In reference to the implementation of the official sensors plugin
// https://github.com/flutter/plugins/tree/master/packages/sensors

enum AudioSource {
  DEFAULT,
  MIC,
  VOICE_UPLINK,
  VOICE_DOWNLINK,
  VOICE_CALL,
  CAMCORDER,
  VOICE_RECOGNITION,
  VOICE_COMMUNICATION,
  REMOTE_SUBMIX,
  UNPROCESSED,
  VOICE_PERFORMANCE
}
enum ChannelConfig { CHANNEL_IN_MONO, CHANNEL_IN_STEREO }
enum AudioFormat { ENCODING_PCM_8BIT, ENCODING_PCM_16BIT }

const AudioSource _DEFAULT_AUDIO_SOURCE = AudioSource.DEFAULT;
const ChannelConfig _DEFAULT_CHANNELS_CONFIG = ChannelConfig.CHANNEL_IN_MONO;
const AudioFormat _DEFAULT_AUDIO_FORMAT = AudioFormat.ENCODING_PCM_8BIT;
const int _DEFAULT_SAMPLE_RATE = 16000;

const int _MIN_SAMPLE_RATE = 1;
const int _MAX_SAMPLE_RATE = 100000;

const EventChannel _microphoneEventChannel =
    EventChannel('aaron.code.com/mic_stream');

Permissions _permission;
Microphone <List <int>> _microphone;

// This function manages the permission and ensures you're allowed to record audio
Future<bool> get permissionStatus async {
  _permission =
      (await Permission.getPermissionsStatus([PermissionName.Microphone]))
          .first;
  if (_permission.permissionStatus != PermissionStatus.allow)
    _permission =
        (await Permission.requestPermissions([PermissionName.Microphone]))
            .first;
  return (_permission.permissionStatus == PermissionStatus.allow);
}

Stream<List<int>> microphone({
  AudioSource audioSource: _DEFAULT_AUDIO_SOURCE,
  int sampleRate: _DEFAULT_SAMPLE_RATE,
  ChannelConfig channelConfig: _DEFAULT_CHANNELS_CONFIG,
  AudioFormat audioFormat: _DEFAULT_AUDIO_FORMAT
}) async* {
  if (_microphone == null)
    _microphone = Microphone(
        audioSource: _DEFAULT_AUDIO_SOURCE,
        sampleRate: _DEFAULT_SAMPLE_RATE,
        channelConfig: _DEFAULT_CHANNELS_CONFIG,
        audioFormat: _DEFAULT_AUDIO_FORMAT);

  yield* _microphone.stream;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:mic_stream/mic_stream.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Stream<List<int>> stream;
  StreamSubscription<List<int>> listener;

  Icon _icon = Icon(Icons.keyboard_voice);
  Color _iconColor = Colors.white;
  Color _bgColor = Colors.cyan;
  bool isRecording = false;

  @override
  void initState() {
    super.initState();
    initPlatformState();

    print("==== Start Example ====");
  }

  void controlMicStream() async {

    if (!isRecording) {

      print("Start Streaming from the microphone...");
      //stream = microphone(audioSource: AudioSource.DEFAULT, sampleRate: 44100, channelConfig: ChannelConfig.CHANNEL_IN_MONO, audioFormat: AudioFormat.ENCODING_PCM_16BIT);
      stream = microphone();
      _updateButton();

      isRecording = true;

      print("Start Listening to the microphone");
      //listener = stream.listen((samples) => samples);
      listener = stream.listen((samples) => print(samples));
    }
    else {
      print("Stop Listening to the microphone");
      listener.cancel();

      isRecording = false;
      print('Stopped listening to the microphone');

      _updateButton();
    }
  }

  void _updateButton() {
    setState(() {
      _bgColor = (isRecording) ? Colors.cyan : Colors.red;
      _icon = (isRecording)  ? Icon(Icons.keyboard_voice) : Icon(Icons.stop);
    });
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    setState(() {});
    if (!mounted) return;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.dark(),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin: mic_stream :: Debug'),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: (){controlMicStream();},
          child: _icon,
          foregroundColor: _iconColor,
          backgroundColor: _bgColor,
        ),
      ),
    );
  }
}


lib/microphone/Microphone.dart
import 'dart:async';

import 'package:flutter/services.dart';

import 'package:mic_stream/mic_stream.dart';

const AudioSource _DEFAULT_AUDIO_SOURCE = AudioSource.DEFAULT;
const ChannelConfig _DEFAULT_CHANNELS_CONFIG = ChannelConfig.CHANNEL_IN_MONO;
const AudioFormat _DEFAULT_AUDIO_FORMAT = AudioFormat.ENCODING_PCM_8BIT;
const int _DEFAULT_SAMPLE_RATE = 16000;

const int _MIN_SAMPLE_RATE = 1;
const int _MAX_SAMPLE_RATE = 100000;

const EventChannel _microphoneEventChannel =
        EventChannel('aaron.code.com/mic_stream');

class Microphone <T> {
    Microphone({
        AudioSource audioSource: _DEFAULT_AUDIO_SOURCE,
        int sampleRate: _DEFAULT_SAMPLE_RATE,
        ChannelConfig channelConfig: _DEFAULT_CHANNELS_CONFIG,
        AudioFormat audioFormat: _DEFAULT_AUDIO_FORMAT
    }) {
        if (sampleRate < _MIN_SAMPLE_RATE || sampleRate > _MAX_SAMPLE_RATE)
            throw (RangeError.range(sampleRate, _MIN_SAMPLE_RATE, _MAX_SAMPLE_RATE));

        _stream = _cast(_squash(_microphoneEventChannel.receiveBroadcastStream([
            audioSource.index,
            sampleRate,
            channelConfig == ChannelConfig.CHANNEL_IN_MONO ? 16 : 12,
            audioFormat == AudioFormat.ENCODING_PCM_8BIT ? 3 : 2
        ]), audioFormat), T.toString());
    }

    Stream <T> _stream;

    Stream <T> get stream => _stream;
    Stream <T> get asBroadcast => _stream.asBroadcastStream();

    Stream <dynamic> _cast (Stream <List <int>> data, String type) async* {
        if (!(await permissionStatus)) throw (PlatformException);

        if (type == "List<int>")
            yield* data;
        if (type == "int")
            yield* data.expand((element) => element);
        throw FormatException("mic_stream: The generic type must be either List<int> or int");
    }

    Stream <List <int>> _squash (Stream <List <int>> data, AudioFormat audioFormat) async* {
        yield* data.map(_squashList);
    }

    List<int> _squashList (List<int> byteSamples) {
        List<int> shortSamples = List();
        bool isFirstElement = true;
        int first;
        for (var sample in byteSamples) {
            if (isFirstElement)
                first = sample;
            else
                shortSamples.add(_squashPair(first, sample));
            isFirstElement = !isFirstElement;
        }
        return shortSamples;
    }

    int _squashPair (int first, int second) {
        return first * 256 + second - 32767;
    }
}
